### PR TITLE
Speed up optimisations

### PR DESCRIPTION
**Pull request by abergeron**
This speeds up the optimisations by about 2X by
 - changing iteration order to avoid going into parts of the graph that will be removed.
 - Making sure we only attempt relevant optimizations for nodes.

This also adds limited support for profiling pipelines to get some insight into what is slow/fast.

We could go a bit further by re-visiting uses of a node when we apply a change on a visited node because that could make the optimizer require even less passes to get a full set done.

### CODE REVIEW

File: myia/opt/opt.py [hunk 1]
```diff
@@ -133,46 +145,144 @@ def deco(f):
...
+                self._d.setdefault(None, []).append(opt)
+                return
+            if not isinstance(ints, tuple):
+                ints = (ints,)
+            for interest in ints:
+                assert isinstance(interest, Primitive)
+                self._d.setdefault(interest, []).append(opt)
+
+        # There could be the option to return do_register also.
+        do_register(opt)
```
[the remark this answers is not in the document]
Comment (abergeron) [reply myia/opt/opt.py]: The idea was to do this and allow do_register to be returned for use as a decorator, but I did not have any use cases.

File: myia/opt/opt.py [hunk 2]
```diff
@@ -133,46 +145,144 @@ def deco(f):
...
-                            tr.replace(node, new)
-                            changes = True
-                            any_changes = True
-                            break
-
-            if not changes:
-                break
-
-        return any_changes
+            mng = manage(graph)  # noqa
```
[the remark this answers is not in the document]
Comment (abergeron) [reply myia/opt/opt.py]: I don't remember and it doesn't seem useful, so I'll remove it.

File: myia/opt/opt.py [hunk 3]
```diff
@@ -133,46 +145,144 @@ def deco(f):
...
+    """Apply a set of local optimizations in bfs order."""
+
+    def __init__(self, node_map, optimizer=None):
+        """Initialize a LocalPassOptimizer."""
+        self.node_map = node_map
         self.optimizer = optimizer
 
     def __call__(self, graph):
-        """Apply optimizations until equilibrium on given graphs."""
+        """Apply optimizations on given graphs in node order."""
```
[the remark this answers is not in the document]
Comment (abergeron) [reply myia/opt/opt.py]: Yes, more or less.  But I want to modify it later to do single-pass full optimisation.

File: myia/opt/opt.py [hunk 4]
```diff
@@ -133,46 +145,144 @@ def deco(f):
...
-        return any_changes
+            mng = manage(graph)  # noqa
+
+        changes = False
+        again = True
+        while again:
+            topo, seen, chg = self.backwards(mng, graph)
+            again = self.forward(mng, topo)
+            changes |= (chg | again)
+            again |= mng.all_nodes.issuperset(seen)
```
[the remark this answers is not in the document]
Comment (abergeron) [reply myia/opt/opt.py]: Well shit.

I knew that this part was mostly a hack, but I did not imagine it was this much of a hack.  I'll drop this condition.

[the remark this answers is not in the document]
Comment (abergeron) [reply myia/opt/opt.py]: Also I've seen cases where it runs more than once, when the gradient is involved.

File: myia/pipeline/steps.py [hunk 5]
```diff
@@ -46,32 +47,66 @@ def __init__(self,
...
+            self.names.append(name)
             self.phases.append(spec)
 
-    def step(self, graph, argspec=None, outspec=None):
+    def step(self, graph, argspec=None, outspec=None, profile=False):
         """Optimize the graph using the given patterns."""
+        if profile:
+            profd = dict()
+            gstart = prof_counter()
+            counter = count(1)
```
[the remark this answers is not in the document]
Comment (abergeron) [reply myia/pipeline/steps.py]: I've tried doing that, but there is so much location-specific code that it quickly becomes really hard to do.

I'm not sure investing a lot of effort into doing that is worthwhile if we are to rewrite all of it in rust anyway.

[the remark this answers is not in the document]
Comment (breuleux) [reply myia/pipeline/steps.py]: We agreed in a meeting that it would be best to rewrite that part in a cleaner way, but that this can be in a future PR.

**Pull request merged**